puppet/modules/jenkins_job_builder/files/theforeman.org/pipelines/release/foreman-x-develop-release.groovy

```diff
@@ -17,51 +17,64 @@ pipeline {
     stages {
         stage('Setup Environment') {
             steps {
-                dir('foreman-installer') {
-                    git(url: 'https://github.com/theforeman/foreman-installer.git', branch: 'develop')
-                    script { commit_hash = archive_git_hash() }
+                script {
+                    if ( env.git_ref?.isAllWhitespace() ) {
+                        error("git_ref parameter is blank")
+                    }
                 }
-                dir('foreman-packaging') { git(url: 'https://github.com/theforeman/foreman-packaging.git', branch: 'rpm/develop') }
-                setup_obal()
-                configureRVM(env.ruby_version)
+                setup_nightly_build_environment(
+                    git_url: git_url,
+                    git_ref: env.git_ref,
+                    project_name: project_name
+                )
             }
         }
-        stage('Build and Archive Source') {
+        stage('Archive Git Commit') {
             steps {
-                dir('foreman-installer') {
-                    withRVM(["bundle install --jobs 5 --retry 5"], env.ruby_version)
-                    withRVM(["bundle exec rake pkg:generate_source"], env.ruby_version)
-                    archiveArtifacts(artifacts: 'pkg/*')
-                    dir('pkg') {
-                        script { sourcefile = sh(script: "ls", returnStdout: true).trim() }
+                script {
+                    dir(project_name) {
+                        commit_hash = archive_git_hash()
                     }
                 }
             }
         }
+        stage('Build and Archive Source') {
+            steps {
+                script {
+                    sourcefile_paths = generate_sourcefiles(project_name: project_name)
```

Comment:
I remember a reason why we previously generated the archive as a separate step: it's not idempotent for the installer. It retrieves the nightly modules from the internet so running it twice can yield different results.

Reply:
Which is ok for the installer since it has no 'merged to master test job' and merges kickoff this job so source is only generated once?

If after going this route, we are still not satisfied with re-building the source then may we need to re-think back to the old strategy but instead of using a Jenkins job as the source of truth, the source is uploaded to a static server for consumption.

Reply:
You're right. Doing this in a single job does mean you get a chicken and egg problem when it's broken. If you need a fix in packaging but also need a recent release artifact for it you can no longer test it in CI. You must do it manually or blindly merge it. Currently the building of a release artifact is one phase and continues as long as the source repository is working. Releasing that into packages is a separate concern. That means our CI can rely on the latest release artifact and test our packaging PRs appropriately. In the new situation we lose this ability. This tight coupling means that any platform issue (EL7, Debian 9, Ubuntu 16.04, Ubuntu 18.04) issue will limit our testing ability. That is my major concern with this design.

I do see that a single pipeline provides a lot more visibility for us humans and would have greatly increased the chance we'd have our recent continuous nightly failure. Because that's more important I'd be OK with proceeding with this, especially because there's another way to regain the testing ability in CI.